proposals/0000-package-versioning.rst

```diff
@@ -0,0 +1,175 @@
+    approach to that. Any version bounds discussion must have the same
+    philosophy if it is to succeed: we need to acknowledge what authors have
+    been willing to do, and what we can reasonably educate people on
+
+    Authors make mistakes with the PVP (I had two package failures in the past
+    week because of it, even though I was blamed for one of them falsely).
+    Automated tooling will make mistakes with bounds. The PVP itself - even if
+    fully followed - does not guarantee 100% success. The only solution that
+    guarantees a 100% success of a build plan is curation (and even that is
+    brittle due to, e.g., different OSes). So we cannot reject a solution
+    because it won't work in some corner case. We need to minimize the
+    probability of failure, accepting that failure will inevitably happen.
+
+
+[1] https://gist.github.com/snoyberg/f6f10cdbea4b9e22d1b83e490ec59a10
+
+
+Motivation
+----------
+
+The problem addressed in this proposal is to reduce the burden on package
+maintainers who are currently required to update the implicit constraints when
+it becomes evident that subsequently published versions of a given dependency
+are indeed compatible with the current package.
+
+Failing to do so can cause the hackage solver to fail, resulting in not being
+able to use particular packages.
+
+Since the problemmatic constraints are essentially outside the control of the
+original developer, some way of moving the management of these out of the hands
+of the package maintainer is proposed.
+
+In Michael Snoyman's words
+
+* Let people continue adding in weird exceptions and cases that they know for
+  certain
+
+* Automate the thing most people are doing, and thereby ensure (1) even people
+  who don't care about the PVP are providing that info, (2) lower bound bumps
+  are not forgotten either, and (3) it's easier to test upper bound relaxes
+  going forward through automated tooling
+
+
+Proposed Change
+---------------
+
+The details of this section will be fleshed out as part of a discussion/dialogue
+with the interested parties.
+
+One possible solution is put forward
+`here <https://github.com/haskell/cabal/issues/3729>`_, but it probably
+requires too large a change to be practicable.
+
+Other solutions involve automatically setting the required implicit constraints
+on upload, using the `--pvp-bounds=both` option when uploading via stack, and a
+similar (to be added) feature for cabal.
+
+Or the implicit bounds could be calculated on hackage as part of the upload
+process, so that maintainers are then only required to specify constraints that
+have to be satisfied, failing which the package cannot be built.
+
+Other solutions ...
+
+The devil is in the details with each of these strategies. We should now being
+constructing concrete proposals around the alternatives, on the way to coming up
+with a single solution that works for all.
```

Comment:
I think the scope of this proposal should be defined better.  Hackage needs a process that actually includes a feedback loop on failure.

A reason why stackage works is that it covers a lot more than what is described here.  

- What happens in failure situations?
    -  In stackage, someone is pinged on github. (feedback).
    -  On hackage, nothing?

- What happens after multiple failures
    - On stackage, it's a breakage of the maintainer's agreement.
    - On hackage, nothing?  In really important cases, an email to haskell-cafe (?) asking to take over a package.

Automating the bounds checking is one thing (and I love stackage), but automating and defining a minimum process, where the minimum includes:
  - There must be a way to contact the maintainer.
  - The maintainer must be notified when matrix.hackage.haskell.org fails.
  - The maintainer must have a way to acknowledge a "ping" from the system.
  - With no signs of life from a maintainer, for a failing package, the package release is marked bad in hackage somehow.

That's just a suggestion, but *there MUST be a feedback loop* and *there must be an escalation system*.  Implicit or explicit.  Hackage could mark packages as bad, emails can be sent, a list can be produced somewhere.   This must exist, and there must be semantics associated with the feedback loop.

Stackage has this, and hackage should also have it.

There is a process around